crates/noirc_evaluator/src/ssa/acir_gen/acir_mem.rs

```diff
@@ -1,22 +1,158 @@
-use crate::ssa::{
-    acir_gen::InternalVar,
-    context::SsaContext,
-    mem::{ArrayId, MemArray},
+use crate::{
+    ssa::{
+        acir_gen::InternalVar,
+        context::SsaContext,
+        mem::{ArrayId, MemArray},
+    },
+    Evaluator,
 };
-use acvm::acir::native_types::Witness;
+use acvm::{
+    acir::{
+        circuit::{directives::Directive, opcodes::Opcode as AcirOpcode},
+        native_types::{Expression, Witness},
+    },
+    FieldElement,
+};
+
 use iter_extended::vecmap;
 use std::collections::BTreeMap;
 
+use super::{
+    constraints::{self, mul_with_witness, subtract},
+    expression_from_witness,
+    operations::{self},
+};
+
+#[derive(Clone, Debug)]
+struct MemItem {
+    op: Expression,
+    value: Expression,
+    index: Expression,
+}
+
 #[derive(Default)]
 pub struct ArrayHeap {
     // maps memory address to InternalVar
     memory_map: BTreeMap<u32, InternalVar>,
+    trace: Vec<MemItem>,
+    staged: BTreeMap<u32, (Expression, Expression)>,
+}
+
+impl ArrayHeap {
+    pub fn commit_staged(&mut self) {
+        for (idx, (value, op)) in &self.staged {
+            let item = MemItem {
+                op: op.clone(),
+                value: value.clone(),
+                index: Expression::from_field(FieldElement::from(*idx as i128)),
+            };
+            self.trace.push(item);
+        }
+        self.staged.clear();
```

Comment:
```suggestion
let items = self.staged.drain(..).map(|(idx, (value, op))| MemItem {
    op: op.clone(),
    value: value.clone(),
    index: Expression::from_field(FieldElement::from(*idx as i128)),
});
self.trace.extend(items);
```

Reply:
I find your suggestion more confusing (it uses drain, map, extend and a lambda) and ends up creating a temporary vector and copying it, instead of a simple loop with push and clear which are straightforward.

Reply:
Hmm I like the iterator solution because it is more specific - drain & map better communicate the intent than a more general for loop. To me it is similar to goto being more general while if, for, and other branching constructs are more specific. The suggestion also does not allocate a temporary vector and is more efficient because it will reserve the needed capacity in self.trace beforehand. I consider this a side benefit though.

Reply:
> Could you explain how it does not create a temporary vector? It seems to me that items is a temporary vector.

Drain returns an iterator rather than a vector, and map as well only works on and returns iterators. Vecs can only be created if `collect` or a similar function that collects internally like vecmap are called. The equivalent code to it would be the same as your previous code but with a `self.trace.reserve(self.staged.len())` line before the for loop.

Reply:
Thank you for your explanation, however staged is a BTreeMap, it is not a vector and does not have the drain() method.

Reply:
Ah, I had assumed since HashMap::drain existed, BTreeMap::drain would as well. We'd have to use `std::mem::take` without it I believe. We can leave this as-is.